pubsub/natsstreaming/nats-streaming.go

```diff
@@ -0,0 +1,239 @@
+	return nil
+}
+
+func (n *natsStreamingPubSub) Publish(req *pubsub.PublishRequest) error {
+	err := n.natStreamingConn.Publish(req.Topic, req.Data)
+	if err != nil {
+		return fmt.Errorf("nats-streaming: error from publish: %s", err)
+	}
+	return nil
+}
+
+func (n *natsStreamingPubSub) Subscribe(req pubsub.SubscribeRequest, handler func(msg *pubsub.NewMessage) error) error {
+	natStreamingsubscriptionOptions, err := n.subscriptionOptions()
+	if err != nil {
+		return fmt.Errorf("nats-streaming: error getting subscription options %s", err)
+	}
+
+	natsMsgHandler := func(natsMsg *stan.Msg) {
+		herr := handler(&pubsub.NewMessage{Topic: req.Topic, Data: natsMsg.Data})
+		if herr != nil {
+		} else {
+			//we only send a successful ACK if there is no error from Dapr runtime
+			natsMsg.Ack()
+		}
+	}
+
+	if n.metadata.subscriptionType == subscriptionTypeTopic {
+		_, err = n.natStreamingConn.Subscribe(req.Topic, natsMsgHandler, natStreamingsubscriptionOptions...)
+	} else if n.metadata.subscriptionType == subscriptionTypeQueueGroup {
+		_, err = n.natStreamingConn.QueueSubscribe(req.Topic, n.metadata.natsQueueGroupName, natsMsgHandler, natStreamingsubscriptionOptions...)
+	}
+	if err != nil {
+		return fmt.Errorf("nats-streaming: subscribe error %s", err)
+	}
+
+	return nil
+}
+
+func (n *natsStreamingPubSub) subscriptionOptions() ([]stan.SubscriptionOption, error) {
+	var options []stan.SubscriptionOption
+
+	if n.metadata.durableSubscriptionName != "" {
+		options = append(options, stan.DurableName(n.metadata.durableSubscriptionName))
+	}
+
+	if n.metadata.deliverNew == deliverNewTrue {
+		options = append(options, stan.StartAt(pb.StartPosition_NewOnly))
+	} else if n.metadata.startAtSequence >= 1 { //messages index start from 1, this is a valid check
+		options = append(options, stan.StartAtSequence(n.metadata.startAtSequence))
+	} else if n.metadata.startWithLastReceived == startWithLastReceivedTrue {
+		options = append(options, stan.StartWithLastReceived())
+	} else if n.metadata.deliverAll == deliverAllTrue {
+		options = append(options, stan.DeliverAllAvailable())
+	} else if n.metadata.startAtTimeDelta > (1 * time.Nanosecond) { //as long as its a valid time.Duration
+		options = append(options, stan.StartAtTimeDelta(n.metadata.startAtTimeDelta))
+	} else if n.metadata.startAtTime != "" {
+		if n.metadata.startAtTimeFormat != "" {
+			startTime, err := time.Parse(n.metadata.startAtTimeFormat, n.metadata.startAtTime)
+
+			if err != nil {
+				return nil, err
+			}
+			options = append(options, stan.StartAtTime(startTime))
+		}
+	}
+
+	//default is auto ACK. switching to manual ACK since processing errors need to be handled
+	options = append(options, stan.SetManualAckMode())
+
+	return options, nil
+}
+
+const inputs = "ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890"
+
+func genRandomString(n int) string {
+	b := make([]byte, n)
+	for i := range b {
+		b[i] = inputs[rand.Int63()%int64(len(inputs))]
+	}
+	clientID := string(b)
+
+	return clientID
+}
```

Comment:
This didn't set random seed so there is a chance to generate the same string. It would use UUID and truncate to 20 characters.

Reply:
added random seed `rand.NewSource(time.Now().UnixNano())`